generic ephemeral volume beta

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

We want to promote generic ephemeral volumes to beta in 1.21.

#### Does this PR introduce a user-facing change?
```release-note
Generic ephemeral volumes are beta.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes
- [Usage]: https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes
```

pohly committed:
generic ephemeral volumes: promote to beta

pohly commented:
/sig storage
/label api-review
/hold

This PR is ready for testing and review, but should not be merged until several other PRs are also merged - see tracking project in https://github.com/orgs/kubernetes-csi/projects/34

ehashman commented:
/retest
/priority important-soon
/triage accepted

pohly commented:
/retest

pohly committed:
generic ephemeral volumes: drop ReadOnly field

As discussed during the alpha review, the ReadOnly field is not really
needed because volume mounts can also be read-only. It's a historical
oddity that can be avoided for generic ephemeral volumes as part
of the promotion to beta.

pohly commented:
/assign @gnufied

Can you review for SIG storage?

pohly committed:
generic ephemeral volume: generate code

This is the result of "make update" minus the testdata update which
will be committed separately.

pohly committed:
testdata: drop the alpha EphemeralVolumeSource.ReadOnly field

This is the result of
   UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test k8s.io/api
after removing the ReadOnly field from the API. The test data
for older releases must be updated because the current code
no longer supports that field.

The removal itself is okay because the struct was declared
as alpha.

Because EphemeralVolumeSource is embedded quite a lot, different types
are affected. Here is one example:

       --- FAIL: TestCompatibility/extensions.v1beta1.DaemonSet/v1.20.0 (0.04s)
            compatibility.go:476: json differs
            compatibility.go:477:   string{
                  	... // 12941 identical bytes
                  	0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x7d, //  |               }|
                  	0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x7d, //  |.              }|
                - 	0x2c, 0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, // -|,.              |
                - 	0x22, 0x72, 0x65, 0x61, 0x64, 0x4f, 0x6e, 0x6c, 0x79, 0x22, 0x3a, 0x20, 0x74, 0x72, 0x75, 0x65, // -|"readOnly": true|
                  	0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x7d, 0x0a, 0x20, //  |.            }. |
                  	0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x7d, 0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, //  |         }.     |
                  	... // 31426 identical bytes
                  }

            compatibility.go:482: yaml differs
            compatibility.go:483:   string{
                  	... // 22893 identical bytes
                  	0x34, 0x22, 0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x65, 0x70, 0x68, 0x65, 0x6d, //  |4".        ephem|
                  	0x65, 0x72, 0x61, 0x6c, 0x3a, 0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, //  |eral:.          |
                - 	0x72, 0x65, 0x61, 0x64, 0x4f, 0x6e, 0x6c, 0x79, 0x3a, 0x20, 0x74, 0x72, 0x75, 0x65, 0x0a, 0x20, // -|readOnly: true. |
                - 	0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,                                           // -|         |
                  	0x76, 0x6f, 0x6c, 0x75, 0x6d, 0x65, 0x43, 0x6c, 0x61, 0x69, 0x6d, 0x54, 0x65, 0x6d, 0x70, 0x6c, //  |volumeClaimTempl|
                  	0x61, 0x74, 0x65, 0x3a, 0x0a, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, //  |ate:.           |
                  	... // 5965 identical bytes
                  }

            compatibility.go:488: proto differs
            compatibility.go:490:   (
                  	"""
                  	... // 459 identical lines
                  	                }
                  	              }
                - 	              2: 1
                  	            }
                  	          }
                  	... // 1083 identical lines
                  	"""
                  )

            compatibility.go:506: wrote expected compatibility data... verify, commit, and rerun tests

liggitt commented:
btw, I opened https://github.com/kubernetes/kubernetes/issues/99716 to track improving the test fixture generator to avoid triggering these enormous changes to unrelated fields in the HEAD fixtures

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1794,11 +1794,6 @@ type EphemeralVolumeSource struct {
-	// Specifies a read-only configuration for the volume.
-	// Defaults to false (read/write).
-	// +optional
-	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
```
comment out rather than delete, and add a tombstone comment so we never reuse this json field name or proto tag. see other "tombstoned" comments in types.go files for examples.

pohly replied:
Done.

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -156,7 +156,7 @@ type VolumeSource struct {
 	// +optional
 	CSI *CSIVolumeSource `json:"csi,omitempty" protobuf:"bytes,28,opt,name=csi"`
-	// Ephemeral represents a volume that is handled by a cluster storage driver (Alpha feature).
+	// Ephemeral represents a volume that is handled by a cluster storage driver.
```
API doc should call out this is a beta feature and depends on the GenericEphemeralVolume feature gate being enabled

pohly replied:
Done.

liggitt reviewed (commented):
have the Alpha → Beta criteria in https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#alpha---beta-graduation been met?

pohly commented:
> have the Alpha → Beta criteria in https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#alpha---beta-graduation been met?

Yes, albeit with some changes still pending in other PRs. Therefore I put this one on hold and will only cancel the hold once they are all in.

pohly commented:
/retest

pohly committed:
generic ephemeral volumes: refresh rbac testdata

This is the result of
  UPDATE_BOOTSTRAP_POLICY_FIXTURE_DATA=true go test k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy

Apparently enabling the GenericEphemeralVolume feature by default
affect this test. The policy that it now tests against is indeed
the one needed for the controller.

pohly commented:
/assign @saad-ali

For confirmation that:
- removal of `EphemeralVolumeSource.ReadOnly` is indeed what he wanted
- this is meant to become beta

pohly commented:
/retest

liggitt commented:
API lgtm, assign once ready for approval

pohly commented:
I'm trying to get @saad-ali's attention for his okay... :sweat_smile: 

FWIW, the only remaining PR why I still have this one here on hold is https://github.com/kubernetes/kubernetes/pull/98918

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

saad-ali commented:
> * removal of EphemeralVolumeSource.ReadOnly is indeed what he wanted

+1 to removal of this ReadOnly field. If we find a volume plugin that doesn't operate correctly without it, we can investigate. Linking https://github.com/kubernetes/kubernetes/issues/70503 for reference

> * this is meant to become beta

Yes the "PVC inline ephemeral volumes (work with CSI driver)" feature is targeted for beta by SIG Storage for the 1.21 release.

/approve
/lgtm

/hold 
To let @liggitt review.

pohly commented:
/assign @liggitt 

For final approval.

Jordan, do you want to hold this until https://github.com/kubernetes/kubernetes/pull/98918 (the PSP validation PR) is merged or can these two PRs proceed independently?

pohly commented:
I added one more small change which removed Saad's LGTM from the PR: the latest commit causes the E2E tests to run in the normal prow jobs.

Jordan, if Saad doesn't get to it, can you also add the LGTM to get this merged?

pohly commented:
I also had to rebase to refresh the `testdata/HEAD/batch.v1.CronJob` files.

pohly commented:
/retest

pohly commented:
/hold cancel

All related PRs have merged, this is the last one for this feature.

liggitt commented:
/approve

for API and client-go changes
storage reviewer has lgtm

pohly commented:
It looks like the 1.21 milestone still needs to be set for this PR, too.

pohly commented:
Rebased again with refreshed `staging/src/k8s.io/api/core/v1/generated.pb.go`.

pohly commented:
/retest

saad-ali reviewed (approved):
/lgtm

pohly commented:
/retest

saad-ali commented:
/milestone v1.21

saad-ali commented:
/lgtm
